### PR TITLE
Add distribution map construction

### PR DESCRIPTION
**Pull request by ThinhinaneIhadadene**
This pull request adds all functions required to construct the distribution map.

### CODE REVIEW

File: include/tiramisu/core.h [hunk 1]
```diff
@@ -2483,7 +2482,38 @@ class computation
...
     computation(std::string name,std::vector<var> iterator_variables, tiramisu::expr e, bool schedule_this_computation, primitive_t t);
 
+    /**
+      * \brief Construct the distribution map of a computation.
+      * Not safe to use. It's currently being implemented.
+      *
+      * A distribution map is a map that describes how the computation is distributed across ranks.
+      * Given the number of available ranks number_of_ranks, the type of the rank rank_type which
+      * is either r_sender or r_receiver, the distribution map will specify for each rank the
+      * iterations it should execute.
```
Comment (rbaghdadi) [inline include/tiramisu/core.h]: Can you specify also how you do the distribution (or partitioning)?
if you have one dimension you will clearly partition that dimension but if you have two dimensions which one will you partition (the outermost, i.e., leftmost)?

File: include/tiramisu/core.h [hunk 2]
```diff
@@ -2483,7 +2482,38 @@ class computation
       */
     computation(std::string name,std::vector<var> iterator_variables, tiramisu::expr e, bool schedule_this_computation, primitive_t t);
 
+    /**
+      * \brief Construct the distribution map of a computation.
+      * Not safe to use. It's currently being implemented.
+      *
+      * A distribution map is a map that describes how the computation is distributed across ranks.
```
Comment (rbaghdadi) [inline include/tiramisu/core.h]: Can you also say that this map "partitions" the computation? (the term partitioning is well known so we should mention it).

File: src/tiramisu_core.cpp [hunk 3]
```diff
@@ -85,6 +85,7 @@ void codegen(const std::vector<tiramisu::buffer *> &arguments, const std::string
 {
     function *fct = global::get_implicit_function();
     fct->codegen(arguments, obj_filename, gen_cuda_stmt);
+
```
Comment (rbaghdadi) [inline src/tiramisu_core.cpp]: Coding style: why did you add a new line?

File: src/tiramisu_core.cpp [hunk 4]
```diff
@@ -7810,4 +7811,98 @@ void tiramisu::buffer::tag_gpu_register() {
     set_auto_allocate(false);
 }
 
+std::string get_rank_string_type(tiramisu::rank_t rank_type)
```
[the remark this answers is not in the document]
Comment (rbaghdadi) [reply src/tiramisu_core.cpp]: I think it is Ok to put here.

File: src/tiramisu_core.cpp [hunk 5]
```diff
@@ -7810,4 +7811,98 @@ void tiramisu::buffer::tag_gpu_register() {
...
+{
+    if (rank_type == rank_t::r_sender)
+        return "r_sender";
```
Comment (rbaghdadi) [inline src/tiramisu_core.cpp]: Can you use shorter names? r_snd and r_rcv? These dimension names will appear a lot in maps and sets so big names will be a problem.

File: src/tiramisu_core.cpp [hunk 6]
```diff
@@ -7810,4 +7811,98 @@ void tiramisu::buffer::tag_gpu_register() {
...
+    else
+        return "r_receiver";
+}
+
+std::vector<std::string> computation::get_static_and_dynamic_dimensions_names()
```
Comment (rbaghdadi) [inline src/tiramisu_core.cpp]: Can you use the function

std::vector<std::string> computation::get_iteration_domain_dimension_names()

instead of creating this one?

The current function fires an error if a name does not exist but I'm ok with you changing it to behave like your new function (i.e., it generates a new name if a name does not exist).

Comment (ThinhinaneIhadadene) [reply src/tiramisu_core.cpp]: I created this one because the existing function returns the names of dimensions of the iteration_domain, not trimmed_time_processor_domain.

Comment (rbaghdadi) [reply src/tiramisu_core.cpp]: Then, can you call something like

computation::get_trimmed_time_space_domain_dimension_names()

This name is close to computation::get_iteration_domain_dimension_names() which we already have. Maybe in the documentation you can talk about the difference between the two functions.
